internal/merkle/proof.go

```diff
@@ -0,0 +1,100 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+package merkle
+
+import (
+	"errors"
+
+	"github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// CreateProofs creates proofs for all the leaves of the binary Merkle tree
+// with the given height. It returns the root hash and the siblings matrix,
+// in bottom-up order.
+//
+// If the number of leaves exceeds the capacity for the given height,
+// an error is returned.
+func CreateProofs(leaves []model.Hash, height uint) (model.Hash, [][]model.Hash, error) {
+	pristineNode := model.Hash{}
+
+	if len(leaves) == 0 {
+		leaves = append(leaves, pristineNode)
+	}
+
+	currentLevel := leaves
+	nextLevel := make([]model.Hash, (len(leaves)+1)/2)
+	siblings := make([][]model.Hash, len(leaves))
+
+	// for each level in the tree, starting from the leaves
+	for levelIdx := 0; levelIdx < int(height); levelIdx++ {
+		calculateSiblings(levelIdx, int(height), currentLevel, siblings, pristineNode)
+		calculateParents(currentLevel, nextLevel, pristineNode)
+
+		aux := currentLevel
+		currentLevel = nextLevel
+		// re-slices current level to avoid allocating multiple arrays
+		nextLevel = aux[:(len(currentLevel)+1)/2]
+		pristineNode = crypto.Keccak256Hash(pristineNode[:], pristineNode[:])
+	}
+
+	// in the end, current level is the root level
+	if len(currentLevel) != 1 {
+		return model.Hash{}, nil, errors.New("too many leaves for height")
+	}
+
+	return currentLevel[0], siblings, nil
+}
+
+// calculateSiblings iterates over each leaf and populates the siblings matrix
+// with the appropriate sibling nodes at the current level.
+//
+// If the level is not full, a pristine node is used instead.
+func calculateSiblings(
+	level, height int,
+	currentLevel []model.Hash,
+	siblings [][]model.Hash,
+	pristineNode model.Hash,
+) {
+	// for each leaf
+	for idx := range siblings {
+		// creates the sibling slice if needed
+		if siblings[idx] == nil {
+			siblings[idx] = make([]model.Hash, height)
+		}
```

Comment:
Why not pre-allocate the whole matrix in `CreateProofs`?

Reply:
Because I would have to loop over the leaves one more time just for this purpose

Reply:
That's true, but it would remove this `if` clause from your code.
On `CreateProofs`, you'd know it is `nil`.
On `CreateSiblings`, you'd like it is not `nil`.

Reply:
Why not use a single allocation? I.e., allocate an array with `len(leaves)*height` entries, and then access it with index (i,j) using i*height+j, where i is the leaf j is the level in the tree. Or you can transpose it, if you want, and access it as j*len(leaves)+i. Doesn't matter if you use the so-called column-major or row-major order. Either way it's a single allocation. It's how people represent matrices in "real life".

Reply:
Done at a764662892a930eda44dd48b1a1278f0da722e41